lib/active_record/connection_adapters/cockroachdb/referential_integrity.rb

```diff
@@ -1,51 +1,36 @@
 # frozen_string_literal: true
-# NOTE(joey): This is cradled from connection_adapters/postgresql/referential_integrity.rb
-# It is commonly used for setting up fixtures during tests.
+
+# The PostgresSQL Adapter's ReferentialIntegrity module can disable and
+# re-enable foreign key constraints by disabling all table triggers. Since
+# triggers are not available in CockroachDB, we have to remove foreign keys and
+# re-add them via the ActiveRecord API.
+#
+# This module is commonly used to load test fixture data without having to worry
+# about the order in which that data is loaded.
 module ActiveRecord
   module ConnectionAdapters
     module CockroachDB
-      module ReferentialIntegrity # :nodoc:
-        def disable_referential_integrity # :nodoc:
-          original_exception = nil
-          fkeys = nil
+      module ReferentialIntegrity
+        def disable_referential_integrity
+          foreign_keys = tables.map { |table| foreign_keys(table) }.flatten
 
-          begin
-            transaction do
-              tables.each do |table_name|
-                fkeys = foreign_keys(table_name)
-                fkeys.each do |fkey|
-                  remove_foreign_key table_name, name: fkey.options[:name]
-                end
-              end
-            end
-          rescue ActiveRecord::ActiveRecordError => e
-            original_exception = e
+          foreign_keys.each do |foreign_key|
+            remove_foreign_key(foreign_key.from_table, name: foreign_key.options[:name])
           end
 
-          begin
-            yield
-          rescue ActiveRecord::InvalidForeignKey => e
-            warn <<-WARNING
-WARNING: Rails was not able to disable referential integrity.
-
-Please go to https://github.com/cockroachdb/activerecord-cockroachdb-adapter
-and report this issue.
-
-    cause: #{original_exception.try(:message)}
-
-              WARNING
-            raise e
-          end
+          yield
```

Comment:
Do you not still want a rescue on the yield? It seems like you still want to re-add the FKs if the block errors.

Reply:
Good question! The `rescue` wouldn't re-add the foreign keys if the `yield` errored. It only captured `ActiveRecord::InvalidForeignKey` errors, and even for those errors the `rescue` re-raised the error causing execution to stop.

The `rescue` was [copied from the PostgreSQL adapter](https://github.com/rails/rails/blob/8bec77cc0f1fd47677a331a64f68c5918efd2ca9/activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb#L18-L31), and it's there to catch permissions problems with disabling foreign keys. Since we're removing foreign keys via the AR API, I don't think we'll run into the same problem. Does that make sense?

Reply:
Makes sense, thanks!